### PR TITLE
feat(update): Add feature to display upgrade notice

### LINKED ISSUE #2789: feat(admin-plugins): display inline notification before updating
## Issue Overview
When Give has a major update, it would be excellent to give our users additional warnings BEFORE they update. This tutorial walks through how that can be done by creating inline update messages:
https://wisdomplugin.com/add-inline-plugin-update-message

Being able to say "Hey, version X.XX has database updates. Please backup your site first. And read all about the awesome features on this blog post" would be really useful.

### PR DESCRIPTION
**Pull request by Sidsector9**
Closes #2789

## Description
On major upgrades like x.0.0 or x.x.0, the upgrade notice will be read from readme.txt

## How Has This Been Tested?
Since we have worked on `release/2.1`, the plugin page will not show if an update is available and also Give doesn't have an `== Upgrade Notice ==` section in its `readme.txt` file. So we are going to spoof a few values to test this.

1. Pull this PR
2. Change current plugin version from 2.1 to 2.0.5
3. Open `admin/plugins.php` file and blindly search-replace the following:

```
Search:  $new_version           = $data['new_version'];
Replace: $new_version           = '3.1.0';

Search:  $current_version_parts = explode( '.', GIVE_VERSION );
Replace: $current_version_parts = explode( '.', '2.9.10' );

Search:  $response = wp_safe_remote_get( 'https://plugins.svn.wordpress.org/give/trunk/readme.txt' );
Replace: $response = wp_safe_remote_get( 'https://plugins.svn.wordpress.org/woocommerce/tags/3.3.5/readme.txt' );

``` 

## Screenshots
There's some UI issue in the plugin listing page as you can see in the image below, WordPress kind of adds the refresh icons at every line, this needs to be worked on:
upgradenotice

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by ravinderk** [comment]
@DevinWalker currently This PR will only show upgrade notice from the plugin but it will not show addon compatibility table in upgrade notice as Woocomerce does.

Let me know if you want to develop that too.

File: give.php [hunk 1]
```diff
@@ -302,8 +302,113 @@ private function init_hooks() {
...
+					continue;
+				}
+
+				$matches = null;
+
+				if ( preg_match( $notice_regexp, $content, $matches ) ) {
+					$notices = (array) preg_split( '~[\r\n]+~', trim( $matches[2] ) );
+
+					if ( version_compare( trim( $matches[1] ), $check_version, '=' ) ) {
+						$upgrade_notice .= '<p class="wc_plugin_upgrade_notice">';
```
Comment (DevinWalker) [inline give.php]: woo prefix... 👎

**Comment by DevinWalker** [comment]
@Sidsector9 for now we just want to display a message like @mathetos describes in the issue. We'll look into enhancing it with add-on compatibility in the future. Please review your code to ensure it's properly updated if you forked some from woo.

**Comment by DevinWalker** [comment]
## Review Findings

I found a few things while reviewing this:

### 1. Number of arguments error

Do you have xDebug on and using latest version 7+ of PHP? If so, you should have seen:

`add_action( 'in_plugin_update_message-Give/give.php', 'give_in_plugin_update_message');`

should be:

`add_action( 'in_plugin_update_message-Give/give.php', 'give_in_plugin_update_message', 10, 2 );`

Without that number of arguments it results in:



## Empty line item

In your screenshot above and in my testing I see an empty line item. Why is that? Please remove it and keep an eye out for UI issues like this:

**Comment by ravinderk** [comment]
@Sidsector9 Did you run code sniffer on your code. Please try to fix the issue as much as you can:

image

**Comment by Sidsector9** [comment]
@DevinWalker I have spoken to @ravinderk about the empty line being added at the end. It happens when I simply echo anything with a `<p>` tag. I have googled to check if this is a WordPress bug but couldn't find much.

**Comment by DevinWalker** [comment]
@Sidsector9 is this also happening within Woo?

**Comment by Sidsector9** [comment]
@DevinWalker  I have fixed the UI so that it looks similar to Woo

upgradenotice

**Comment by DevinWalker** [comment]
Looks good, thanks @Sidsector9

**Pull request merged**